packages/counters/init.lua

```diff
@@ -42,9 +42,14 @@ function package.formatCounter (_, counter)
 end
 
 function package:formatMultilevelCounter (counter, options)
-  local maxlevel = options and options.level or #counter.value
-  local minlevel = options and options.minlevel or 1
+  local opts = options or {}
```

Comment:
```suggestion
  options = options or {}
```

... and rename `opts` → `options` throughout. Unless there is a reason to use a different variable throughout I'd rather just use the table passed rather than a copy with a similar name.

Reply:
> Re tests: I would suggest Busted based unit tests rather than full page layout regression tests since we're after the math here more than the resulting typesetting.

That makes perfect sense. I'll work on it, then, and update the PR.

> Re class scope: (...)
Yep, points noted. For the record about my "asymmetry" concern, I now have code (after replacing the corresponding deprecated `SILE.xxx` previous calls) such as:
```
local n = self.class.packages.counters:formatMultilevelCounter(
   self.class:getMultilevelCounter(counter)
)
```
I would have _kind of_ expected, either:
```
local n = self.class:formatMultilevelCounter(
   self.class:getMultilevelCounter(counter)
)
```
Or
```
local n = self.class.packages.counters:formatMultilevelCounter(
   self.class.packages.counter:getMultilevelCounter("mycounter")
)
```
With _probably_ a preference for the second form - The first makes it sound, at least at first glance, that it could be harder for one to use another "counter" package (compatible at API level, but perhaps with additional options and methods, etc.) instead of the stock one.

BTW, this does raise another possible idea: to accept either a counter object _or_ a counter id in methods such as `formatMultilevelCounter`, so the above could become, without loosing any current functionality:

```
local n = self.class.packages.counters:formatMultilevelCounter("mycounter")
```
I'm not sure how such methods accepting different types for an argument are regarded here, though. (Developers' tastes sometimes differ on such an approach)

Reply:
I'd like to move *away* from functions and methods that accept multiple types. They will be the hardest parts to substitute in Rust or Teal or whatever in the future. I know there are still a lot of use cases for this and some of them are even quite useful. The gray area here is where type casting comes into play. Any type enforcement we implement is going to have to account for the cases where we can cast to the right type. For example anything coming from SIL input is going to be a string, but depending on the command we cast it no numbers or lengths or whatever. So the long term rule will be "only accept type X or anything that can be cast to X". Lua doesn't handle this very gratefully, but Rust for example has features that can be used to automatically handle that kind of situational conversion. The difference is it is backwards from Lua: the function doesn't test for a string and then try to convert it to the right type, the string would have to have a trait that enabled it to be cast to something else.

I'm kind of torn on this. I actually typed up my fist answer as "I suggest adding a similar method that accepts an id, fetches the counter, then passes it to the current method such as `:formatMultilevelCounterId()`", but given all the existing precedent I wouldn't object if you thought just implementing it a test for string vs. counter object and used strings to fetch counters by ID.